akka-typed/src/main/scala/akka/typed/Extensions.scala

```diff
@@ -0,0 +1,62 @@
+/**
+ * Copyright (C) 2017 Lightbend Inc. <http://www.lightbend.com>
+ */
+package akka.typed
+
+/**
+ * Marker trait for extensions
+ *
+ * @see [[ExtensionId]]
+ */
+trait Extension
+
+/**
+ * Identifier and factory for an extension.
+ *
+ * {{{
+ *   class MyExtension(system: ActorSystem[_]) {
+ *     ...
+ *   }
+ *
+ *   object MyExtension extends ExtensionId[MyExtension] {
+ *     override def create(system: ActorSystem[_]): MyExtension =
+ *       new MyExtension(system)
+ *   }
+ * }}}
+ *
+ * @tparam T The concrete extension type
+ */
+abstract class ExtensionId[T <: Extension] {
+
+  def createExtension(system: ActorSystem[_]): T
+
+  final def apply(system: ActorSystem[_]): T = system.registerExtension(this)
+  final def create(system: ActorSystem[_]): T = apply(system)
```

Comment:
I think `get` is a better name, even though it's lazy get or create.

There is probably some scaladoc you can steal also